streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -2097,276 +2102,113 @@ <VO, VR> KStream<K, VR> outerJoin(final KStream<K, VO> otherStream,
-     * For this case, all data of the stream will be redistributed through the repartitioning topic by writing all
-     * records to it, and rereading all records from it, such that the join input {@code KStream} is partitioned
-     * correctly on its key.
+     * Join records of this stream with {@link KTable}'s records using non-windowed inner equi join.
+     * In contrast to {@link #join(KTable, ValueJoiner)}, but only if the used {@link KTable} is backed by a
+     * {@link org.apache.kafka.streams.state.VersionedKeyValueStore VersionedKeyValueStore}, the additional
+     * {@link Joined} parameter allows to specify a join grace-period, to handle out-of-order data gracefully.
      *
-     * @param table  the {@link KTable} to be joined with this stream
-     * @param joiner a {@link ValueJoiner} that computes the join result for a pair of matching records
-     * @param joined      a {@link Joined} instance that defines the serdes to
-     *                    be used to serialize/deserialize inputs of the joined streams
-     * @param <VT>   the value type of the table
-     * @param <VR>   the value type of the result stream
-     * @return a {@code KStream} that contains join-records for each key and values computed by the given
-     * {@link ValueJoiner}, one for each matched record-pair with the same key
-     * @see #leftJoin(KTable, ValueJoiner, Joined)
-     * @see #join(GlobalKTable, KeyValueMapper, ValueJoiner)
+     * <p>For details about stream-table semantics, including co-partitioning requirements, (auto-)repartitioning,
+     * and more see {@link #join(KTable, ValueJoiner)}.
+     * If you specify a grace-period to handle out-of-order data, see further details below.
+     *
+     * <p>To handle out-of-order records, the input {@link KTable} must use a
+     * {@link org.apache.kafka.streams.state.VersionedKeyValueStore VersionedKeyValueStore} (specified via a
+     * {@link Materialized} parameter when the {@link KTable} is created), and a join
+     * {@link Joined#withGracePeriod(Duration) grace-period} must be specified.
+     * For this case, {@code KStream} records are buffered until the end of the grace period and the {@link KTable}
+     * lookup is performed with some delay.
+     * Given that the {@link KTable} state is versioned, the lookup can use "event time", allowing out-of-order
+     * {@code KStream} records, to join to the right (older) version of a {@link KTable} record with the same key.
+     * Also, {@link KTable} out-of-order updates are handled correctly by the versioned state store.
+     * Note, that using a join grace-period introduces the notion of "late records", i.e., records with a timestamp
+     * smaller than the defined grace-period allows; these "late records" will be dropped, and not join computation
+     * is triggered.
+     * Using a versioned state store for the {@link KTable} also implies that the defined "history retention" provides
```

Comment:
It's a bit unclear what you mean by history retention here. It's quoted and not mentioned before, not explained.